File: src/test/java/com/fifo/ticketing/domain/user/service/EmailAuthServiceTests.java
package com.fifo.ticketing.domain.user.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import jakarta.mail.MessagingException;
import jakarta.mail.internet.MimeMessage;
import jakarta.servlet.http.HttpSession;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.mail.javamail.JavaMailSender;
import org.thymeleaf.context.Context;
import org.thymeleaf.spring6.SpringTemplateEngine;

class EmailAuthServiceTests {

    @InjectMocks
    private EmailAuthService emailAuthService;

    @Mock
    private RedisService redisService;

    @Mock
    private JavaMailSender mailSender;

    @Mock
    private SpringTemplateEngine templateEngine;

    @Mock
    private MimeMessage mimeMessage;

    @Mock
    private HttpSession session;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);

        when(mailSender.createMimeMessage()).thenReturn(mimeMessage);
        when(templateEngine.process(eq("mail"), any(Context.class))).thenReturn(
            "<html><body>Code: 123456</body></html>");
    }

    @Test
    void testSendEmail() throws MessagingException {
        String toMail = "test@test.com";

        emailAuthService.sendEmail(toMail);

        verify(redisService, times(1)).setValuesWithTimeout(
            eq("EAC:" + toMail), anyString(), eq(5 * 60 * 1000L));
        verify(mailSender, times(1)).send(mimeMessage);
    }

    @Test
    void testCorrectAuthCode() {
        String email = "test@test.com";
        String authCode = "123456";
        when(redisService.getValues("EAC:" + email)).thenReturn(authCode);

        boolean checked = emailAuthService.checkAuthCode(email, authCode, session);

        verify(redisService).deleteValues("EAC:" + email);
        verify(session).setAttribute("emailVerified", email);
        assertThat(checked).isTrue();
    }

    @Test
    void testIncorrectAuthCode() {
        String email = "test@test.com";
        String authCode = "123456";
        when(redisService.getValues("EAC:" + email)).thenReturn("654321");

        boolean checked = emailAuthService.checkAuthCode(email, authCode, session);

        verify(redisService, never()).deleteValues(any());
        verify(session, never()).setAttribute(any(), any());
        assertThat(checked).isFalse();
    }

}

File: src/test/java/com/fifo/ticketing/domain/user/service/EmailAuthServiceTests.kt
package com.fifo.ticketing.domain.user.service

import com.fifo.ticketing.global.service.RedisService
import io.mockk.*
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import jakarta.mail.Message
import jakarta.mail.MessagingException
import jakarta.mail.internet.MimeMessage
import jakarta.servlet.http.HttpSession
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.mail.javamail.JavaMailSender
import org.thymeleaf.context.Context
import org.thymeleaf.spring6.SpringTemplateEngine

@ExtendWith(MockKExtension::class)
class EmailAuthServiceTests {
    private val TEST_EMAIL_FORM = "test@example.com"

    @MockK
    private lateinit var redisService: RedisService

    @MockK
    private lateinit var mailSender: JavaMailSender

    @MockK
    private lateinit var templateEngine: SpringTemplateEngine

    @MockK
    private lateinit var mimeMessage: MimeMessage

    @MockK
    private lateinit var session: HttpSession

    lateinit var emailAuthService: EmailAuthService

    @BeforeEach
    fun setUp() {
        MockKAnnotations.init(this)

        every { mailSender.createMimeMessage() } returns mimeMessage
        every {
            templateEngine.process(eq("mail"), match { it is Context })
        } returns "<html><body>Code: 123456</body></html>"

        every { mimeMessage.setSubject(any()) } just Runs
        every { mimeMessage.setText(any()) } just Runs
        every { mimeMessage.setText(any(), any(), any()) } just Runs
        every { mimeMessage.setContent(any(), any()) } just Runs
        every { mimeMessage.setFrom(any<String>()) } just Runs
        every { mimeMessage.addRecipients(eq(Message.RecipientType.TO), any<String>()) } just Runs

        emailAuthService = EmailAuthService(
            redisService,
            mailSender,
            templateEngine,
            TEST_EMAIL_FORM
        )
    }

    @Test
    @Throws(MessagingException::class)
    fun testSendEmail() {
        val toMail = "test@test.com"

        every {
            redisService.setValuesWithTimeout(
                "EAC:$toMail", any(), 5 * 60 * 1000L
            )
        } just Runs

        every { mailSender.send(mimeMessage) } just Runs

        emailAuthService.sendEmail(toMail)

        verify(exactly = 1) {
            redisService.setValuesWithTimeout("EAC:$toMail", any(), 5 * 60 * 1000L)
        }

        verify(exactly = 1) { mailSender.send(mimeMessage) }
    }

    @Test
    fun testCorrectAuthCode() {
        val email = "test@test.com"
        val authCode = "123456"
        every { redisService.getValues("EAC:$email") } returns authCode
        every { redisService.deleteValues("EAC:$email") } just Runs
        every { session.setAttribute("emailVerified", email) } just Runs

        val checked = emailAuthService.checkAuthCode(email, authCode, session)

        assertThat(checked).isTrue()
        verify { redisService.deleteValues("EAC:$email") }
        verify { session.setAttribute("emailVerified", email) }
    }

    @Test
    fun testIncorrectAuthCode() {
        val email = "test@test.com"
        val authCode = "123456"
        every { redisService.getValues("EAC:$email") } returns "654321"

        val checked = emailAuthService.checkAuthCode(email, authCode, session)

        assertThat(checked).isFalse()
        verify(exactly = 0) { redisService.deleteValues(any()) }
        verify(exactly = 0) { session.setAttribute(any(), any()) }
    }
}